### ISSUE TITLE
Expression evaluator fails when multiple conditions are specified without paranthesis

### ISSUE THREAD

**Issue by vikasrohit** [appirio-tech/connect-app#2793]
This `( details.appDefinition.deliverables contains \'design\' || details.appDefinition.deliverables contains \'dev-qa\' )` does not work while this
`( (details.appDefinition.deliverables contains \'design\') || (details.appDefinition.deliverables contains \'dev-qa\') )` works

fyi @maxceem @RishiRajSahu

**Comment by maxceem** [comment]
@vikasrohit I think the reason why this `( details.appDefinition.deliverables contains \'design\' || details.appDefinition.deliverables contains \'dev-qa\' )` doesn't work, is because our implementation supports very basic operator precedence which doesn't covers combination of `contains` and `||`.

While we probably can patch it now. We don't have some general approach here so during the time logic would become more complex and trickier for support.

There is some existent implementation of shunting algorithm in JS https://github.com/Calamari/shunting-yard.js

The implementation is very as simple as ours. But it supports the precedence of operators which can be configured in very descriptive way. It looks like it supports adding custom operators and functions so we can easily add what we need. We can define precedence for all operators we use using the standard approach in JS https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence.

The biggest benefit to move to this lib, as it has a framework which would help us to keep the implementation clean when we add new operators or functions. But implementation is still very simple so if necessary we can easily fix it if needed.

If we decide to move, we will need to configure operators/functions we currently use, as that implementation supports just few operators out of the box.

@RishiRajSahu as you are more in context of this feature, maybe you have some thought regarding this.

**Comment by vikasrohit** [comment]
> because our implementation supports very basic operator precedence which doesn't covers combination of `contains` and `||`.

**Comment by vikasrohit** [comment]
>If we decide to move, we will need to configure operators/functions we currently use, as that implementation supports just few operators out of the box.

What exact operators we are missing, we are using right now, in this out of box features?

**Comment by maxceem** [comment]
So far out of the box it supports only math operators:`+`, `-`, `*`, `/` and `^`. Also, it supports parenthesis `(` and `)` out of the box.  

While we actively use `||`, `&&`, `==`, `!=`, `contains`, `hasLength`. And we also support but I didn't see us using `>`, `<`, `!`.

Just have noticed about this lib: `Beware that your tokens don't contain whitespaces at the beginning or the end, so "+" is not the same as " +".`. So I guess if we switch we will have to fix it to be more relaxed on spaces.

So there are tradeoffs:
- Our code:
   - (+) can fix now quicker
   - (-) with time every fix would be harder and pron to issues
   - (-) most likely will need to refactor anyway to something like that lib to keep code easy to support

- Switch to lib
   - (-) need to spend time to switch
   - (-) not sure if there any drawbacks in the code which we will need to fix
   - (+) easy to add new operators later and easy to support
   - (+) the lib code is very simillar to ours and same easy as ours, so we can fix it if need

**Comment by vikasrohit** [comment]
Do we have any other library as option?

If we don't, we can clone this library in our org and start customizing it. We also need to publish this to our npm repo as well so that we can use it across applications e.g. we might need this evaluator in near future for project service as well.

**Comment by RishiRajSahu** [comment]
@maxceem I think we just have to update the logic of `hasPrecedence` method to keep our existing parser work.
and to highlight - 
>(-) with time every fix would be harder and prone to issues

 I think once we correctly update the `hasPrecedence` method by defining the precedence of operators with numeric values we are good, whenever we need to add a new custom operator we just have to correctly define the precedence of that new operator and this effort will be there as well for the lib which you shared

> (-) most likely will need to refactor anyway to something like that lib to keep code easy to support

yeah I am more aligned towards that and that's what I meant by defining numeric precedence for ops and we can follow definitely [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence)

**Comment by maxceem** [comment]
> Do we have any other library as option?

I think I found another one, but it was huge - too much for our needs.

> I think we just have to update the logic of `hasPrecedence` method to keep our existing parser work.

Yeah, can do like that.

**Comment by RishiRajSahu** [comment]
@maxceem Pushed change for precedence logic https://github.com/appirio-tech/connect-app/commit/6419ae9fd698f779344db5a6e6106ac4128ce159, should we create another issue for our CF to add test case by community which could break our parser (assuming every condition should have balanced parentheses) ?

**Comment by maxceem** [comment]
@RishiRajSahu you mean to ask the community to find the ways to break our parser?

**Comment by RishiRajSahu** [comment]
@maxceem yeah, I am looking for some way to find the breaking test cases and then make our parser capable for that.

**Comment by vikasrohit** [comment]
@RishiRajSahu that is an interesting idea, we can do that. As we are able to get it fixed for now, lets keep our custom parser and get it tested via community. Let them write unit tests that break it.
And in long term, lets plan for a better library upgrade. May be the one @maxceem suggested. I would try to fork it in our repo and publish to npm. Once that is done, we can configure that as library in our eco system and can use it across applications.

**Comment by maxceem** [comment]
> And in long term, lets plan for a better library upgrade. May be the one @maxceem suggested. I would try to fork it in our repo and publish to npm. Once that is done, we can configure that as library in our eco system and can use it across applications.

Btw, I think the upgrade done by Rishi solves the issues at least for our needs for some time. The other library now has only one advantage: it's has kind of framework structure. But feature wise I think now we are more or less equal.
Now we can also easily add new operators in our implementation.

**Comment by vikasrohit** [comment]
Agreed. Lets keep an eye on library for this work as that might save some future work.

**Comment by maxceem** [comment]
@RishiRajSahu I've create a new issues for creating tests cases https://github.com/appirio-tech/connect-app/issues/2855

While this issues is fixed and already in DEV.

**Comment by RishiRajSahu** [comment]
@maxceem Thanks for tracking the remains in another issue.
